components/usage/pkg/db/types.go

```diff
@@ -0,0 +1,59 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package db
+
+import (
+	"database/sql/driver"
+	"errors"
+	"fmt"
+	"github.com/relvacode/iso8601"
+	"time"
+)
+
+func NewVarcharTime(t time.Time) VarcharTime {
+	return VarcharTime(t.UTC())
+}
+
+func NewVarcharTimeFromStr(s string) (VarcharTime, error) {
+	parsed, err := iso8601.ParseString(string(s))
+	if err != nil {
+		return VarcharTime{}, fmt.Errorf("failed to parse as ISO 8601: %w", err)
+	}
+	return VarcharTime(parsed), nil
+}
+
+// VarcharTime exists for cases where records are inserted into the DB as VARCHAR but actually contain a timestamp which is time.RFC3339
+type VarcharTime time.Time
+
+// Scan implements the Scanner interface.
+func (n *VarcharTime) Scan(value interface{}) error {
+	if value == nil {
+		return fmt.Errorf("nil value")
+	}
+
+	switch s := value.(type) {
+	case []uint8:
+		if len(s) == 0 {
+			return errors.New("failed to parse empty varchar time")
+		}
+
+		parsed, err := iso8601.ParseString(string(s))
+		if err != nil {
+			return fmt.Errorf("failed to parse %v into ISO8601: %w", string(s), err)
+		}
+		*n = VarcharTime(parsed.UTC())
+		return nil
+	}
+	return fmt.Errorf("unknown scan value for VarcharTime with value: %v", value)
+}
+
+// Value implements the driver Valuer interface.
+func (n VarcharTime) Value() (driver.Value, error) {
+	return time.Time(n).UTC().Format(time.RFC3339Nano), nil
+}
+
+func (n VarcharTime) String() string {
+	return time.Time(n).Format(time.RFC3339Nano)
```

Comment:
This is a different format to what we currently store in the DB (`2006-01-02T15:04:05.999999999Z07:00` vs. `2006-01-02T15:04:05.999Z`) and wouldn't parse using JS `new Date(<string>)`.

Reply:
String() is only used for printing it out in a human readable form. It's the interface to `Stringer()` which printers use to serialize a struct. Given that the format used by JS is fairly non-standard, the human readable format here is the standard RFC3339.

There may be gotchas in other places with timestamps, and I'll be adding much more in regards to tests for those. For now, we're focused on reading records only so writes (while important) are not as a big of a deal at this point.

Reply:
A secondary problem with what we store right now is that we abuse VarChar to store timestamps. We rely on somewhat obscure feature of MySQL which allows for VarChart timestamp comparisons to work. The same, for example, is not true in PostgreSQL

Reply:
> String() is only used for printing it out in a human readable form

@easyCZ Despite what format is standardized where, we should really thrive to have _one_ format everywhere. Everything else will give us headaches when doing cross-service deugging.

Reply:
> A secondary problem with what we store right now is that we abuse VarChar to store timestamps

While I admin that this solution disturbed my personally feeling of "this is not the way it should be done (sic)" I think it worked rather well.

> We rely on somewhat obscure feature of MySQL which allows for VarChart timestamp comparisons to work

I think it's regular string comparison ordering, and not that far fetched. :wink: 

>  The same, for example, is not true in PostgreSQL

True, but we would not translate the current scheme 1:1 to postgresql, anyway. Even if we were to do it today. :upside_down_face:

Reply:
Usage of ISO 8601 is pretty non standard in Go. I'd strongly recommend that for human consumption, we stick with the standard way.

IMO, this is a non-issue in the bigger picture of this PR. But if it enables approval of this PR and allows us to progress usage based billing, I'll happily change it.

Reply:
As promised, a follow-up to this issue is in a PR now: https://github.com/gitpod-io/gitpod/pull/10490 (with tests)